keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+get hands on feedback from developers to guide our design decisions around
+asthetics/egronomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+}
+
+func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+```
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. There are a couple reasons for this:
+
+- If a client has multiple code paths where it makes apply requests to the same
+  object, but with different field sets, they must use different field manager
+  names. If they use the same field manager name they will cause fields to be
+  accidentally removed or disowned. This is a potential foot gun we would like to
+  avoid.
+
+- Apply requests always conflict with update requests, even if they were made by
+  the same client with the same field manager name. This is by design. So when a
+  controller migrates from update to apply, it will need to deal with conflicts
+  regardless of what field manager name is used.
+
+### Generated apply configuration types
+
+All fields present in an apply configuration become owned by the applier after
+when the apply request succeeds. Go structs contain zero valued fields which are
+included even if the user never explicitly sets the field. Required boolean
+fields are a good example of fields that would be applied incorrectly using go
+structs, e.g. `ContainerStatus.Ready` (required, not omitempty). Because of this
+we cannot use the existing go structs to represent apply configurations.
+
+<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
+Finalize which alternative to use based on developer feedback. See the
+[Alternatives](#alternatives) for a complete list, but are currently focusing on
+the two below alternatives. We are working with the Kubebuilder community to
+gather feedback on what developers prefer.
+<<[/UNRESOLVED]>>
+
+#### Alternative 1: Genreated structs where all fields are pointers
+
+Example usage:
+
+```go
+&appsv1apply.Deployment{
+  Name: &pointer.StringPtr("nginx-deployment"),
```

Comment:
I don't think you wanted to have a pointer to a pointer here (you can't take a pointer of an r-value).

Reply:
+1

Reply:
Oops, this was a typo. The existing pointer package in Kubernetes that I'm assuming would be used here correctly returns a pointer so this wouldn't have compiled. Fixed.